src/cryptonote_core/blockchain.cpp

```diff
@@ -4193,10 +4199,66 @@ bool Blockchain::handle_block_to_main_chain(const block& bl, const crypto::hash&
     goto leave;
   }
 
+  // verify all non-input consensus rules for txs inside the pool supplement (if not inside checkpoint zone)
+#if defined(PER_BLOCK_CHECKPOINT)
+  if (!fast_check)
+#endif
+  {
+    tx_verification_context tvc{};
+    // If fail non-input consensus rule checking...
+    if (!ver_non_input_consensus(extra_block_txs, tvc, hf_version))
+    {
+      MERROR_VER("Pool supplement provided for block with id: " << id << " failed to pass validation");
+      bvc.m_verifivation_failed = true;
+      goto leave;
+    }
+  }
+
   size_t coinbase_weight = get_transaction_weight(bl.miner_tx);
   size_t cumulative_block_weight = coinbase_weight;
 
   std::vector<std::pair<transaction, blobdata>> txs;
+  //                          txid     weight mempool?
+  std::vector<std::tuple<crypto::hash, size_t, bool>> txs_meta;
+
+  // This will be the data sent to the ZMQ pool listeners for txs which skipped the mempool
+  std::vector<txpool_event> txpool_events;
+
+  // this lambda returns relevant txs back to the mempool
+  auto return_txs_to_pool = [this, &txs, &txs_meta, &hf_version]()
+  {
+    if (txs_meta.size() != txs.size())
+    {
+      MERROR("BUG: txs_meta and txs not matching size!!!");
+      return;
+    }
+
+    for (size_t i = 0; i < txs.size(); ++i)
+    {
+      // if this transaction wasn't ever in the pool, don't return it back to the pool
+      const bool found_in_pool = std::get<2>(txs_meta[i]);
+      if (!found_in_pool)
+        continue;
+
+      transaction &tx = txs[i].first;
+      const crypto::hash &txid = std::get<0>(txs_meta[i]);
+      const blobdata &tx_blob = txs[i].second;
+      const size_t tx_weight = std::get<1>(txs_meta[i]);
+
+      // We assume that if they were in a block, the transactions are already known to the network
+      // as a whole. However, if we had mined that block, that might not be always true. Unlikely
+      // though, and always relaying these again might cause a spike of traffic as many nodes
+      // re-relay all the transactions in a popped block when a reorg happens. You might notice that
+      // we also set the "nic_verified_hf_version" paramater. Since we know we took this transaction
+      // from the mempool earlier in this function call, when the mempool has the same current fork
+      // version, we can return it without re-verifying the consensus rules on it.
+      cryptonote::tx_verification_context tvc{};
+      if (!m_tx_pool.add_tx(tx, txid, tx_blob, tx_weight, tvc, relay_method::block, true,
+          hf_version, hf_version))
+        MERROR("Failed to return taken transaction with hash: " << txid << " to tx_pool");
+    }
```

Comment:
Looks like we should `notify_txpool_event` each newly added tx in this for loop as well.

Sequence of events:
1) Receive new fluffy tx as arg.
2) Still missing some txs in the fluffy block.
3) We end up in this loop, adding the new fluffy txs to the pool.

The fluffy txs from 3 will never get pub'd.

It looks like the only circumstance where we might receive *some* fluffy txs in a fluffy block but are still missing txs is from a node running different code. But I figure we'd want to pub txs in that case anyway as these would be newly valid txs entering the pool.

Reply:
Fluffy txs aren't added to the mempool on block verification failure, which means the ZMQ notification happens later: https://github.com/monero-project/monero/blob/fb3adce341518e06d1862657109d16976923a88b/src/cryptonote_core/blockchain.cpp#L4238. However, maybe this shouldn't be the behavior. Perhaps we *do* add them to the mempool and perform a ZMQ notify right here?

Reply:
Whoops, that was a miss on my part. I actually prefer it as is, since we don't expect nodes to include some fluffy txs while missing others. Adding the fluffy txs to the pool seems like unnecessary optimization.